src/meili/match_route.cc

```diff
@@ -208,69 +216,145 @@ bool MergeRoute(std::vector<EdgeSegment>& route, const State& source, const Stat
+                 state = mapmatcher.state_container().state(match.stateid);
 
       // get the route between the two states by walking edge labels backwards
       // then reverse merge the segments together which are on the same edge so we have a
       // minimum number of segments. in this case we could at minimum end up with 1 segment
-      std::vector<EdgeSegment> segments;
-      MergeRoute(segments, prev_state, state);
+      segments.clear();
+      if (!MergeRoute(prev_state, state, segments) && !route.empty()) {
+        route.back().discontinuity = true;
+        prev_idx = curr_idx;
+        prev_match = &match;
+        continue;
+      }
 
-      // TODO remove: the code is pretty mature we dont need this check its wasted cpu
-      if (!ValidateRoute(mapmatcher.graphreader(), segments.begin(), segments.end(), tile)) {
-        throw std::runtime_error("Found invalid route");
+      // trivial routes (both states are the same point) will have no path
+      // this seems to be an optimization from routing.cc
+      if (segments.empty())
+        continue;
+
+      // we need to cut segments where a match result is marked as a break
+      new_segments.clear();
+      cut_segments(match_results, prev_idx, curr_idx, segments, new_segments);
+
+      // have to merge route's last segment and segments' first segment together if its not a
+      // discontinuity or a break
+      if (!prev_match->is_break_point && !route.empty() && !route.back().discontinuity &&
+          route.back().edgeid == new_segments.front().edgeid) {
+
+        EdgeSegment& first_half = route.back();
+        EdgeSegment& second_half = new_segments.front();
+
+        // Before merge:               -1          prev_idx      prev_idx, last_match_idx
+        //                first_half    ------------------>      ------------------>  second_half
+        // After merge :               -1                                  last_match_idx
+        //                             ------------------------------------------->  second_half
+        if (first_half.first_match_idx == -1) {
+          second_half.source = 0.f;
+          second_half.first_match_idx = -1;
+        }
+        // Before merge:            first_match_idx, prev_idx   prev_idx,        -1
+        //                first_half    ------------------>      ------------------>  second_half
+        // After merge :             first_match_idx                             -1
+        //                              ------------------------------------------->  second_half
+        if (second_half.last_match_idx == -1) {
+          second_half.target = 1.f;
+          second_half.first_match_idx = route.back().first_match_idx;
+        }
+        route.pop_back();
       }
 
+      // debug builds check that the route is valid
+      assert(ValidateRoute(mapmatcher.graphreader(), new_segments.begin(), new_segments.end(), tile));
```

Comment:
for years we had a TODO about removing this but seems its ok to put this into an assert